docs/proposals/device-crd.md

```diff
@@ -0,0 +1,207 @@
+Device management is a key feature required for an edge computing platform.
+This proposal addresses how can we manage devices from the cloud and synchronize
+the device updates between edge nodes and cloud.
+
+### Goals
+
+Device management must:
+* provide APIs for managing devices from the cloud.
+* synchronize the device updates between cloud and edge nodes.
+
+### Proposal
+We propose using Kubernetes Custom Resource Definitions (CRDs) to describe device metadata/status and a custom controller to sync device metadata/status updates
+between the edge node and the cloud.
+
+### Syncing Device Updates From Cloud To Edge
+<img src="../images/device-crd/device-updates-cloud-edge.png">
+The device controller watches device updates in the cloud and relays them to the edge node. These updates are stored locally by the device twin . The mapper gets these updates via the broker and operates on the device based on the updates.
+
+### Syncing Device Updates From Edge To Cloud
+<img src="../images/device-crd/device-updates-edge-cloud.png">
+The mapper watches devices for updates and reports them to the event bus via the broker. The event bus reports the actual state to the device twin which stores it locally and then syncs the updates to the cloud. The device controller watches for device updates from the edge ( via the cloudhub ) and updates the actual state in the cloud via the device CRD APIs.
+
+### Modelling Devices as CRD’s
+<img src="../images/device-crd/device-crd-model.png">
+
+[Custom resources](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) allows us to define declarative APIs for custom resources. Devices attached to edge nodes can be described using a custom resource and managed using APIs installed in a Kubernetes cluster. Users can interact with the device CRD’s using standard `kubectl` operations or using client libraries. The Device controller is a CRD controller whose job is to ensure the device updates are synced between the cloud and the edge node.
+
+### Device Custom Resource Overview
+The device custom resource layout is shown below. It contains standard fields ( Name, Namespace, Labels , Annotations etc. ) which help the Kubernetes ecosystem to manage resources. In addition, it contains custom fields in the Spec and Status detailed further below.
+<img src="../images/device-crd/device-crd.png">
+
+The `Spec` contains the following fields
+
+| Field                 | Description                                                                      |
+|-----------------------|----------------------------------------------------------------------------------|
+| Attributes            | A list of device model attributes. Each device attribute describes the data type ,<br />permissible values, defaults etc.|
+| Attribute Visitors    | The attribute visitors describe how to access the attribute values. <br />Each attribute has a visitor. |
```

Comment:
visitor describes the internal details of how to access a particular attribute for a given device.
for example , for a [CC2650 Sensor Tag](http://processors.wiki.ti.com/index.php/CC2650_SensorTag_User%27s_Guide),  if we want to read the temperature value, we need to access a specific register with a uuid. In addition, many more details could be needed. The primary consumer of this would be the mapper which would connect to the Sensor Tag and read the data from a BLE service and charactersitic. For other protocols , the way to access the data could vary.